rbx_binary/src/core.rs

```diff
@@ -1,3 +1,42 @@
+use std::io::{self, Read, Write};
+
+use rbx_dom_weak::RbxValue;
+
 pub static FILE_MAGIC_HEADER: &[u8] = b"<roblox!";
 pub static FILE_SIGNATURE: &[u8] = b"\x89\xff\x0d\x0a\x1a\x0a";
-pub const FILE_VERSION: u16 = 0;
\ No newline at end of file
+pub const FILE_VERSION: u16 = 0;
+
+pub trait BinaryType<T: ?Sized> {
+    fn read_binary<R: Read>(
+        source: &mut R,
+    ) -> io::Result<RbxValue>;
+
+    fn write_binary<W: Write>(
+        output: &mut W,
+        value: &T,
+    ) -> io::Result<()>;
+
+    fn read_array<R: Read>(
+        source: &mut R,
+        count: usize,
+    ) -> io::Result<Vec<RbxValue>> {
+        let mut result = Vec::new();
+
+        for _ in 0..count {
+            result.push(Self::read_binary(source)?);
+        }
+
+        Ok(result)
+    }
+
+    fn write_array<W: Write>(
+        output: &mut W,
+        values: &[&T]
+    ) -> io::Result<()> {
+        for value in values {
+            Self::write_binary(output, value)?;
+        }
+
+        Ok(())
+    }
```

Comment:
Can we collapse this trait to just the `read_array` and `write_array` methods? I'd rather that we have to consider the loop for every type instead of having to deal with the boilerplate of writing both versions of the method in cases where they don't overlap.

We can provide the `read_single` and `write_single` methods with provided implementations that use the `read_array` and `write_array` forms instead and specialize those if we need to!

Additionally, when we write `read_array` in each case, we should use [`Vec::with_capacity`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.with_capacity), which will save us a ton of allocations since we know our size in advance!

I'm also wondering if the write interface should accept an iterator instead of a slice, which lets us avoid intermediate allocations when we're using derived values!